### ISSUE TITLE
AMOEBA stretch-bend: OpenMM does not handle the "general case"

### ISSUE THREAD

**Issue by jayponder** [pandegroup/openmm#800]
I think I’ve discovered an issue in the OpemMM implementation of AMOEBA’s stretch-bend term. A typical parameter, taken from one of the TINKER parameter files, would be:

strbnd        2    1    3       7.20       7.20

This is a stretch-bend term for the angle consisting of atom classes 2, 1 and 3, connected as the angle 2-1-3. The stretch-bend energy for this angle is a sum of two terms. The first is a force constant
times the 2-1 bond length deviation times the 2-1-3 angle deviation. The second term is a force constant times the 1-3 bond deviation times the 2-1-3 angle deviation (where “deviation” means difference from the ideal/target value for the bond length or angle).

In the above parameter, the first 7.20 is the force constant for the first term, i.e., the one containing the 2-1 bond deviation. The second 7.20 is the force constant for the term involving the 1-3 bond length.
In many case, the two force constants are often the same (and in the AMOEBA Bio09 and Pro13 parameter sets, they are always the same). However, there are many other molecules where the two force constants in a single stretch-bend parameter are different. OpenMM appears to not deal with the case where the two force constants are, in fact, different. It seems to use the first force canstant value (I think..) for both of the terms arising from a single stretch-bend occurrence.

In future AMOEBA parameterizations, we envision that the two force constants in each stretch-bend will be allowed to move to different values. So while this is not significant for users of the current AMOEBA biomolecule parameters, it is for users of the present general organic parameters (AMOEBA09) and will be a frequent/general issue going forward.

**Comment by jchodera** [comment]
From what I can tell, it does appear that the current API in fact only allows *one* force constant to be specified for each stretch-bend term:
https://github.com/pandegroup/openmm/blob/master/plugins/amoeba/openmmapi/include/openmm/AmoebaStretchBendForce.h#L65-L78
```C++
    /**
     * Add a stretch-bend term to the force field.
     *
     * @param particle1     the index of the first particle connected by the stretch-bend
     * @param particle2     the index of the second particle connected by the stretch-bend
     * @param particle3     the index of the third particle connected by the stretch-bend
     * @param lengthAB      the equilibrium length of the stretch-bend in bond ab [particle1, particle2], measured in nm
     * @param lengthCB      the equilibrium length of the stretch-bend in bond cb [particle3, particle2], measured in nm
     * @param angle         the equilibrium angle in radians
     * @param k             the force constant for the stretch-bend
     * @return the index of the stretch-bend that was added
     */
    int addStretchBend(int particle1, int particle2, int particle3, double lengthAB,  double lengthCB, double angle,
                       double k );
```

**Comment by jayponder** [comment]
Hi,

Yes, I found the same code. I’m fairly certain the current OpenMM
implementation just needs to be generalized to get both force constants,
and then use them for the two separate terms in each stretch-bend.

In the past (and in the current amoebabio09 and amoebapro13 parameters)
we have usually constrained both of the force constants to be the same
when fitting parameters. But we have discovered that this (unnecessary)
restriction is not a particularly good idea in many cases. In some of
the current amoeba09 organic parameters, and generally going forward,
the two stretch-bend constants will be allowed to float independently.

Thanks, Jay

> On Feb 6, 2015, at 11:11 PM, John Chodera <notifications@github.com> wrote:
> 
> From what I can tell, it does appear that the current API in fact only allows one force constant to be specified for each stretch-bend term:
> https://github.com/pandegroup/openmm/blob/master/plugins/amoeba/openmmapi/include/openmm/AmoebaStretchBendForce.h#L65-L78
> 
>     /**
>      * Add a stretch-bend term to the force field.
>      *
>      * @param particle1     the index of the first particle connected by the stretch-bend
>      * @param particle2     the index of the second particle connected by the stretch-bend
>      * @param particle3     the index of the third particle connected by the stretch-bend
>      * @param lengthAB      the equilibrium length of the stretch-bend in bond ab [particle1, particle2], measured in nm
>      * @param lengthCB      the equilibrium length of the stretch-bend in bond cb [particle3, particle2], measured in nm
>      * @param angle         the equilibrium angle in radians
>      * @param k             the force constant for the stretch-bend
>      * @return the index of the stretch-bend that was added
>      */
> 
>     
> int addStretchBend(int particle1, int particle2, int particle3, double lengthAB,  double lengthCB, double
>  angle,
>                        
> double k );
> —
> Reply to this email directly or view it on GitHub.

--
Jay W. Ponder                          Phone:  1-314-935 4275
Chemistry, Campus Box 1134             Fax:    1-314-935 4481
Washington University in St. Louis
One Brookings Drive                    Email:  ponder@dasher.wustl.edu
St. Louis, Missouri 63130  USA         WWW:    http://dasher.wustl.edu/

**Comment by swails** [comment]
I actually noticed this same thing myself a week ago, or so, when I was adding OpenMM-Amoeba support to my Python modeling library.  I know @peastman is busy with triclinic box support, so I'll try my hand at adding this flexibility.

**Comment by swails** [comment]
This should be fixed by #801, so this issue can be closed.

Thanks!
